tasks.py

```diff
@@ -138,19 +158,21 @@ def queue_daemon():
     except ImportError:
         pass
 
-    while 1:
+    stopper = threading.Event()
+    worker = ProcessMessage(stopper, 'worker')
 
-        msg = redis.blpop(REDIS_QUEUE_KEY)
-        func, key, args, kwargs = loads(msg[1])
-        try:
-            func(*args, **kwargs)
-        except Exception as e:
-            tb = traceback.format_exc()
-            logger.info(tb)
-            client.captureException()
+    def signalHandler(signum, frame):
+        stopper.set()
+        sys.exit(0)
 
-        if redis.llen(REDIS_QUEUE_KEY) == 0:
-            logger.info("Hurrah! Done merging Metro PDFs.")
+    signal.signal(signal.SIGINT, signalHandler)
+    signal.signal(signal.SIGTERM, signalHandler)
+    # Custom Timeout error: 2 minutes.
+    signal.signal(signal.SIGALRM, timeout_handler)
+    signal.alarm(120)
```

Comment:
So, the way I read this is that you want this to die no matter what after 2 minutes. Is that the intent?

Reply:
I'd like `makePacket` to timeout after two minutes, if unoconv freezes, which happens from time-to-time.

I cannot create a signal in `makePacket`, since it throws this error: 
```
ValueError: signal only works in main thread
```

Is there a way to associate the custom Timeout with the makePacket function?

Reply:
What does it look like when it freezes? Does it seem like something is still happening? Or something died and didn't get cleaned up? Or something else? I am curious mainly because it seems like that might be a better way of approaching that problem: figuring out how to handle what is going on with that particular subprocess call. Instead of just setting some kind of arbitrary timeout.

That said, when you `join` a thread, you can give it an optional timeout. What that does is it will wait for the thread to finish what it's doing or get to the timeout (whatever comes first)

Reply:
Locally, LibreOffice stays open (i.e., I can see the icon in my dock), and all activity ceases in the metro-merger worker. Closing LibreOffice fixes the problem, and the workers carries on. So, perhaps, if the process stops for, say, 2 minutes, we could force quit LibreOffice...e.g., `pkill soffice.bin`....though I am not sure if killing all Office processes is wise. (In any case, processing a file in `makePacket` should not take more than two minutes: if it does, something happened on the LibreOffice side.)

Reply:
Huh, have you tried giving the [subprocess call](https://docs.python.org/3/library/subprocess.html#subprocess.call) a timeout?

Reply:
Ah, okay. That's smart. I'll try to implement this.

Reply:
Okay...I added a timeout to the subprocesses, e.g., `call(['unoconv', '-f', 'pdf', filename], timeout=120)`

However, it looks like the subprocess runs fine: it's a [LibreOffice, not unoconv problem](https://github.com/dagwieers/unoconv/issues/342). 

I see two options:

(1) force quit LibreOffice, after two minutes
(2) kill the worker, after two minutes (via your suggestion of join)